api/v1alpha1/circuitrelay_types.go

```diff
@@ -0,0 +1,104 @@
+/*
+Copyright 2022.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"github.com/libp2p/go-libp2p-core/peer"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	ma "github.com/multiformats/go-multiaddr"
+)
+
+// This is intended to mimic peer.AddrInfo
+type AddrInfoBasicType struct {
+	ID       string        `json:"id"`
+	Addrs    []string      `json:"addrs"`
+	addrInfo peer.AddrInfo `json:"-"`
+}
+
+func (a *AddrInfoBasicType) Parse() error {
+	id, err := peer.IDB58Decode(a.ID)
+	if err != nil {
+		return err
+	}
+	addrs := make([]ma.Multiaddr, len(a.Addrs))
+	for i, addr := range a.Addrs {
+		maddr, err := ma.NewMultiaddr(addr)
+		if err != nil {
+			return err
+		}
+		addrs[i] = maddr
+	}
+	ai := peer.AddrInfo{
+		ID:    id,
+		Addrs: addrs,
+	}
+	a.addrInfo = ai
+	return nil
+}
+
+func (a *AddrInfoBasicType) AddrInfo() *peer.AddrInfo {
```

Comment:
Following the code, wouldn't returning a pointer to `a.addrInfo` allow the calling function to simply perform changes onto `a.addrInfo` anyway?
And if so, then wouldn't this be the same as exporting the `a.AddrInfo` field in the first place?

I'm not trying to be a heavy code critic, just wondering whether this is something Go-specific that I'm not understanding.

Reply:
This is true, we could just reference the field directly.

The reason for not including it is json serialization. If it's capitalized, the json encoder complains.